spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/core/Mqttv5ClientManager.java

```diff
@@ -0,0 +1,153 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.mqtt.core;
+
+import org.eclipse.paho.mqttv5.client.IMqttAsyncClient;
+import org.eclipse.paho.mqttv5.client.IMqttToken;
+import org.eclipse.paho.mqttv5.client.MqttAsyncClient;
+import org.eclipse.paho.mqttv5.client.MqttCallback;
+import org.eclipse.paho.mqttv5.client.MqttConnectionOptions;
+import org.eclipse.paho.mqttv5.client.MqttDisconnectResponse;
+import org.eclipse.paho.mqttv5.common.MqttException;
+import org.eclipse.paho.mqttv5.common.MqttMessage;
+import org.eclipse.paho.mqttv5.common.packet.MqttProperties;
+
+import org.springframework.integration.mqtt.event.MqttConnectionFailedEvent;
+import org.springframework.util.Assert;
+
+/**
+ * @author Artem Vozhdayenko
+ * @since 6.0
+ */
+public class Mqttv5ClientManager extends AbstractMqttClientManager<IMqttAsyncClient, MqttConnectionOptions>
+		implements MqttCallback {
+
+	private final MqttConnectionOptions connectionOptions;
+
+	public Mqttv5ClientManager(String url, String clientId) {
+		this(buildDefaultConnectionOptions(url), clientId);
+	}
+
+	public Mqttv5ClientManager(MqttConnectionOptions connectionOptions, String clientId) {
+		super(clientId);
+		Assert.notNull(connectionOptions, "'connectionOptions' is required");
+		this.connectionOptions = connectionOptions;
+		if (!this.connectionOptions.isAutomaticReconnect()) {
+			logger.info("If this `ClientManager` is used from message-driven channel adapters, " +
+					"it is recommended to set 'automaticReconnect' MQTT connection option. " +
+					"Otherwise connection check and reconnect should be done manually.");
+		}
+		Assert.notEmpty(connectionOptions.getServerURIs(), "'serverURIs' must be provided in the 'MqttConnectionOptions'");
+		setUrl(connectionOptions.getServerURIs()[0]);
+	}
+
+	private static MqttConnectionOptions buildDefaultConnectionOptions(String url) {
+		Assert.notNull(url, "'url' is required");
+		var connectionOptions = new MqttConnectionOptions();
+		connectionOptions.setServerURIs(new String[]{ url });
+		connectionOptions.setAutomaticReconnect(true);
+		return connectionOptions;
+	}
+
+	@Override
+	public synchronized void start() {
+		if (getClient() == null) {
+			try {
+				var client = new MqttAsyncClient(getUrl(), getClientId());
+				client.setManualAcks(isManualAcks());
+				client.setCallback(this);
+				setClient(client);
+			}
+			catch (MqttException e) {
+				throw new IllegalStateException("could not start client manager", e);
+			}
+		}
+		try {
+			getClient().connect(this.connectionOptions)
+					.waitForCompletion(this.connectionOptions.getConnectionTimeout());
+		}
+		catch (MqttException e) {
+			logger.error("could not start client manager, client_id=" + getClientId(), e);
+
+			var applicationEventPublisher = getApplicationEventPublisher();
```

Comment:
Why no reconnect logic as we discussed?
Exactly the same what we have right now in the `Mqttv5PahoMessageDrivenChannelAdapter.doStart()`.
That exactly was a purpose of our discussion around that `setAutomaticReconnect(true)` option.

Reply:
Sorry, I don't see any reason to call `reconnect` manually because if automatic reconnect is enabled, the client is reconnected automatically anyway (tests show it works). This `reconnect` just zeroes out the timer.
Also if automatic reconnect is disabled, I thought we have discussed that the user gets info message about it and have to subscribe to `MqttConnectionFailedEvent`

Reply:
Ok I see now, I was really sure that there was no point for reconnect but this makes sense now. Indeed I should have made git blame, sometimes I think that smth might be implemented for a reason but not this time, haha.